src/gasPriceOracle/oracle.ts

```diff
@@ -0,0 +1,95 @@
+import { BigNumber, providers } from "ethers";
+import { toBN } from "../utils";
+
+export type GasPriceEstimate = {
+  maxFeePerGas: BigNumber;
+  maxPriorityFeePerGas?: BigNumber;
+};
+
+type ValidatedFeeData = {
+  gasPrice: BigNumber;
+  maxFeePerGas: BigNumber;
+  maxPriorityFeePerGas: BigNumber;
+};
+
+interface GasPriceFeed {
+  (provider: providers.Provider): Promise<GasPriceEstimate>;
+}
+
+function validateFeeData(val: BigNumber | null): BigNumber {
+  return BigNumber.isBigNumber(val) ? val : toBN(-1);
+}
```

Comment:
Instead of validating each parameter, why not just have a function that validates the entire struct?

Something like

```ts
type EIP1559FeeData = {
  maxFeePerGas: BigNumber;
  maxPriorityFeePerGas: BigNumber;
};

type LegacyFeeData = {
  gasPrice: BigNumber;
};

function isEip1559FeeData(input: providers.FeeData): input is EIP1559FeeData {
  return feeData.maxFeePerGas !== undefined && 
              BigNumber.isBigNumber(feeData.maxFeePerGas)) &&
              feeData.maxPriorityFeePerGas !== undefined &&
              BigNumber.isBigNumber(feeData.maxPriorityFeePerGas));
}

function isLegacyFeeData(input: providers.FeeData): input is LegacyFeeData {
  return feeData.gasPrice !== undefined && 
              BigNumber.isBigNumber(feeData.gasPrice));
}

function sendError(provider: providers.Provider, message: string): never {
  throw new Error("error message here");
}

function getEip1559FeeData(provider: providers.Provider): EIP1559FeeData {
  const feeData = await provider.getFeeData();
  if (!isEip1559FeeData(feeData)) sendError(provider, "Expected EIP1559 fee data");
  return feeData;
}

function getLegacyFeeData(provider: providers.Provider): LegacyFeeData {
  const feeData = await provider.getFeeData();
  if (!isLegacyFeeData(feeData)) sendError(provider, "Expected Legacy fee data");
  return feeData;
}

...
```

Reply:
A few notes on this.

`function isEip1559FeeData(input: providers.FeeData): input is EIP1559FeeData {` is a [type predicate](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates). When returning true, this forces typescript to understand that the input conforms to the EIP1559FeeData type.

`function sendError(provider: providers.Provider, message: string): never {` this never tells typescript that the function can never return, i.e. it will always throw. So it won't be considered in control flow analysis in the calling function.

`if (!isEip1559FeeData(feeData)) sendError(provider, "Expected EIP1559 fee data");` After this check, typescript should be convinced that the feeData object conforms to our EIP1559FeeData type.

Reply:
> Instead of validating each parameter, why not just have a function that validates the entire struct?
> 
> Something like
> 
> ```ts
> type EIP1559FeeData = {
>   maxFeePerGas: BigNumber;
>   maxPriorityFeePerGas: BigNumber;
> };
> 
> type LegacyFeeData = {
>   gasPrice: BigNumber;
> };
> ```

One question: Do you think it's necessary to have two different types for this?

I was hoping to consolidate underneath an EIP-1559 definition of `{ maxFeePerGas, maxPriorityFeePerGas }`, with the hope that if the caller is producing a type 0 transaction, they can just take the value from maxFeePerGas anyway. Aside from the naming difference, maxFeePerGas maps pretty well onto gasPrice anyway. In this way we could (hopefully) abstract away the transaction type information, so that callers of this don't need to consider the differences.

Reply:
No, I think you've effectively routed around this problem!